### ISSUE TITLE
Perf issues with animation

### ISSUE THREAD

**Issue by TheSeamau5** [evancz/elm-effects#5]
Here, as promised, the animation example: https://github.com/TheSeamau5/elm-components/tree/master/examples/5

Admittedly the example is quite contrived (rotating 100 squares at once). But, this serves to showcase the issue with how animations are done in elm-components. While the encapsulation is great, this means that each square gets updated one at a time instead of all at ones.

The slow version is called `SpinSquareList` and the fast version is called `SpinSquareListManual`. The difference between the two is quite jarring and is noticeable starting from around 20 squares or so...

**Comment by rgrempel** [comment]
Do you still have the 100 rotating squares example around? The original link seems to be broken.

I'm curious whether the Ticker approach here would make any difference:

https://github.com/rgrempel/elm-architecture-tutorial/blob/plain-old-tasks/elm-tick/Ticker.elm

It would actually at least consolidate the tick requests, so if that's where the overhead is, it could help.

Also, I wonder whether the overhead actually lies in repeatedly calling the 'view' method, rather than 'update'. In principle, we don't really care about what 'view' outputs until the next animation frame is needed, so it might be nice to somehow avoid recalculating it until then. I haven't really thought through how to do that, though.

**Comment by rgrempel** [comment]
I implemented my own "100 rotating squares" as a SpinSquareList, and I saw the stuttering originally reported -- it's not subtle!

My elm-ticker code consolidates all 100 requests for an animation frame (i.e. it only calls requestAnimationFrame() rather than 100 times). But it turns out that, by itself, doesn't help -- it's still just as bad.

However, I worked up some code with the logic I mentioned above -- that is, once we request an animation frame, filter out view updates so the view method doesn't run until the animation frame is requested and all the update methods have run. (Since there's no point running our view logic for intermediate states).

That does the trick -- the animation is smooth, even at 200 rotating squares. It starts stuttering again around 350, but that's much better, anyway.

I'll work on cleaning up the code and pushing it to Github, but in the meantime I just wanted to let you know that the issue is substantially improved if you can avoid calling the view logic multiple times while waiting for an animation frame to be needed.

**Comment by rgrempel** [comment]
Here is the code which I mentioned:

https://github.com/rgrempel/elm-ticker

To see the difference which the view optimization makes, compare rotating 100 squares with and without it:

* [Unoptimized](https://rgrempel.github.io/elm-ticker/unoptimized.html)

* [Optimized](https://rgrempel.github.io/elm-ticker/optimized.html)

**Comment by evancz** [comment]
> My elm-ticker code consolidates all 100 requests for an animation frame (i.e. it only calls requestAnimationFrame() once rather than 100 times). But it turns out that, by itself, doesn't help -- it's still just as bad.

elm-effects is already doing this.

> However, I worked up some code with the logic I mentioned above -- that is, once we request an animation frame, filter out view updates so the view method doesn't run until the animation frame is requested and all the update methods have run. (Since there's no point running our view logic for intermediate states).

Is the following summary of this idea correct? "Apply all the animation updates in one big batch before doing any `view` stuff."

If so, I think this would be a relatively simple change to the existing library. Instead of triggering a bunch of messages from the `requestAnimationFrame` we just trigger one that is a list of all the necessary ticks. We tag that in some nice way and send it to the core runner of stuff. In there, we do a `foldl update model tickList` to apply all the changes in one go. That way the `view` function is only called once.

Is that plausible?

**Comment by rgrempel** [comment]
elm-effects tries to consolidate tick requests, but it only consolidates tick requests which are made during a **single call** to `update` -- or, thinking of it another way, a single 'pulse' through the signal graph. So, using typical Elm idioms, nothing is actually consolidated -- since each call to `update` gets dispatched to some specific component anyway. (For instance, in example number 8, nothing actually gets consolidated).

Whereas one of the things rgrempel/elm-ticker adds is that it consolidates **all** tick requests until requestAnimationFrame actually fires. So, even across multiple calls to `update`, requestAnimationFrame is only ever called once, until it actually fires.

But, as I mentioned above, this turns out (by itself) not to make a ton of difference in practice.

The other thing which rgrempel/elm-ticker facilitates is more or less equivalent to "Apply all the animation updates in one big batch before doing any view stuff" -- that turns out to make a huge difference to animation performance. 

However, I think you are under-estimating what is necessary in order to accomplish that. I don't think you can easily modify elm-effects to do this: "Instead of triggering a bunch of messages from the requestAnimationFrame we just trigger one that is a list of all the necessary ticks." The problem, again, is that "all the necessary ticks" is "every tick requested until requestAnimationFrame actually fires" -- not "every tick requested during a single run of `update`". And elm-effects currently tracks the latter, not the former.

In other words, in order to achieve the view optimization, you actually need to track a bit of isolated state across multiple calls to `update` -- that's the clever bit of what rgrempel/elm-ticker does -- that is, it tracks some isolated state without exposing much complexity to the client.

**Comment by evancz** [comment]
That all sounds fine to me. Why not just make the change here though? I don't see how it would change the public API at all. You are essentially saying "we should do batching better". I agree.

I think @jwmerrill has more expertise on this topic than me, since he worked out how to do it for normal Elm rendering. It's quite tricky, so if we need to do it, it'd be cool if we could just do it once with a patch change.

**Comment by rgrempel** [comment]
I guess the best way I can say it is that it seems to me that innovation in this area ought to be possible outside of the elm-effects package. It would be a symptom of a problem of some kind if that were not the case.

Remember that all I'm actually asking for (in one way or another) is access to the browser's `requestAnimationFrame` function, because I think I can do some interesting things with it. Is it really necessary for me to buy into how elm-effects does things in order to get that?

I mean, if it turns out that I've done something useful, and it turns out that elm-effects can be modified to do the same useful thing, then great -- that's how innovation happens. But it shouldn't be mandatory for me to build on elm-effects.

In fact, part of what I'm trying to investigate with rgrempel/elm-ticker is that whether we can do interesting effects-related work in de-centralized packages, without having to standardize on a single `Effects` type, in a single package. I think there are large benefits to be gained if that turns out to be true. Now, is it true? Who knows -- but I'd like to find out.

**Comment by evancz** [comment]
Yes, you are saying things that I agree with. I started this interaction saying [this](https://github.com/evancz/elm-effects/pull/13#issuecomment-140213854):

> I get that things are not going as fast as you want, and I apologize for that. I am focusing my efforts on stuff that will unblock all this kind of stuff, not just your particular scenario that you have right now.

Yes, it'll be cool when lots of folks can experiment with these ideas. There are other concerns that make it more complex than it sounds and I'm working on it. Please answer the questions I am asking you.

> Why not just make the change here though? I don't see how it would change the public API at all.

You are the one who said [the goal was to fix #5](https://github.com/evancz/elm-effects/pull/13#issuecomment-140216176). Is that not what you want to do anymore? I don't have time for long talks, I do have time to help get a fix for this through.

P.S. on the question of "can effects be generic so people can plug in what they like?" Yes. I know. I had a draft API that did this. It was fancier, so I did not release it in the initial version. There are many things to balance here. This is one that is solved and also one that is not a pressing concern yet. The value of "there's a way of doing things in Elm and it's super simple" is very very high, so I'm not interested in muddying the messaging to get a thing that (a) folks writing frontends are not asking for right now and (b) I know how to do already.

**Comment by rgrempel** [comment]
Ah, my apologies, I suppose that I didn't mean that my goal was to literally fix #5. Instead, it was more to demonstrate how the problem identified in #5 could be dealt with using an alternative approach.

On the larger question, I think you're doing an astonishingly good job of managing the development of the Elm community. So, if that means that rgrempel/elm-ticker doesn't appear in the Elm package repository, that's fine with me.

**Comment by evancz** [comment]
On the note of how to be generic, the core method is to make the `Effect` type have a hole for whatever you want.

```elm
type Effects fx = ...
```

Where the user can put whatever message they want in there as `fx`. You can then map over that part to make it fancier. The only big change is that `start` needs to take a function like `List fx -> Task x ()` which describes how to run a batch of user defined effects.

**Comment by evancz** [comment]
So I originally had a module like `Effects.Special` for the special effects ;) I had implemented the simpler API in terms of it. I didn't want all that out in the initial release.

If this is something you are interested in and passionate about, it may make sense to start exploring this kind of thing. Figure out how to do stuff like GraphQL with it (shouldn't be tooo crazy) or how to hook up to custom data bases. This all would be good.

I apologize for being curt before, I think I get extra stressed about anything that touches on native libraries because I know we can do better but it's still not great now.

**Comment by rgrempel** [comment]
No problem -- you're trying to do something difficult and amibitious -- and 99% of the time you're very patient with people. I'll take a closer look at an `Effects.Special` approach when I get a chance.

**Comment by evancz** [comment]
I wrote up a more complete version of this [here](https://gist.github.com/evancz/e4e4173382a9024f95d1) and shared it on the mailing list.
